File: packages/aretro/src/aretro/exec/read_evidence_detail.py
"""Read one selected value from an aretro evidence payload artifact."""

from __future__ import annotations

from pathlib import Path
from typing import Annotated, Any, NoReturn, cast

import click

from asdl_core.clinkr.ensure import Ensure
from asdl_core.clinkr.exit import ClinkrExit
from asdl_core.clinkr.failure import ClinkrFailure
from asdl_core.clinkr.models import ClinkrModel
from asdl_core.clinkr.operation import clinkr_operation
from asdl_core.payloads.errors import PayloadError
from asdl_core.payloads.lookup import read_json_payload_artifact, resolve_json_pointer


class ReadEvidenceDetailRequest(ClinkrModel):
    payload_path: Annotated[
        str,
        click.Option(["--payload-path"], type=click.STRING, required=True),
    ]
    json_pointer: Annotated[
        str,
        click.Option(["--json-pointer"], type=click.STRING, required=True),
    ]


class ReadEvidenceDetailResult(ClinkrModel):
    payload_path: str
    json_pointer: str
    value: Any


@clinkr_operation(
    name="read-evidence-detail",
    help="Read one selected value from an aretro evidence payload artifact.",
)
def run_read_evidence_detail(
    ctx: click.Context,
    request: ReadEvidenceDetailRequest,
) -> ClinkrExit[ReadEvidenceDetailResult]:
    del ctx
    _validate_data_pointer(request.json_pointer)
    envelope = _read_raw_payload_envelope(Path(request.payload_path))
    _validate_success_envelope(envelope, payload_path=request.payload_path)
    _validate_payload_data(envelope, payload_path=request.payload_path)
    value = _resolve_detail_value(envelope, request.json_pointer)
    return ClinkrExit.ok(
        ReadEvidenceDetailResult(
            payload_path=request.payload_path,
            json_pointer=request.json_pointer,
            value=value,
        )
    )


def _validate_data_pointer(pointer: str) -> None:
    if pointer == "/data" or pointer.startswith("/data/"):
        return
    Ensure.fail(
        error_type="invalid_request",
        message=f"JSON Pointer must target the payload data document under /data: {pointer!r}",
    )


def _read_raw_payload_envelope(payload_path: Path) -> dict[str, Any]:
    try:
        document = read_json_payload_artifact(payload_path, allowed_roles=frozenset({"raw"}))
    except PayloadError as error:
        _raise_clinkr_failure_for_payload_error(error)

    if not isinstance(document, dict):
        Ensure.fail(
            error_type="payload_lookup_failed",
            message=f"Raw payload artifact must contain a Clinkr envelope object: {payload_path}",
        )
    return cast(dict[str, Any], document)


def _validate_success_envelope(envelope: dict[str, Any], *, payload_path: str) -> None:
    exit_code = envelope.get("exit_code")
    if not isinstance(exit_code, int) or isinstance(exit_code, bool) or exit_code != 0:
        Ensure.fail(
            error_type="payload_lookup_failed",
            message=f"Raw payload artifact must be a successful Clinkr envelope: {payload_path}",
        )
    if "data" not in envelope:
        Ensure.fail(
            error_type="payload_lookup_failed",
            message=f"Raw payload artifact is missing Clinkr data: {payload_path}",
        )


def _validate_payload_data(envelope: dict[str, Any], *, payload_path: str) -> None:
    data = envelope.get("data")
    if not isinstance(data, dict):
        Ensure.fail(
            error_type="payload_lookup_failed",
            message=f"Raw payload artifact data must be an aretro detail object: {payload_path}",
        )
    schema_version = data.get("schema_version")
    if not isinstance(schema_version, int) or isinstance(schema_version, bool):
        Ensure.fail(
            error_type="payload_lookup_failed",
            message=(
                f"Raw payload artifact data is missing a supported schema version: {payload_path}"
            ),
        )
    if schema_version != 1:
        Ensure.fail(
            error_type="payload_lookup_failed",
            message=f"Raw payload artifact schema version is unsupported: {payload_path}",
        )


def _resolve_detail_value(envelope: dict[str, Any], pointer: str) -> Any:
    try:
        return resolve_json_pointer(envelope, pointer)
    except PayloadError as error:
        _raise_clinkr_failure_for_payload_error(error)


def _raise_clinkr_failure_for_payload_error(error: PayloadError) -> NoReturn:
    raise ClinkrFailure(error_type=error.error_type, message=error.message) from error

File: ts/packages/aretro/src/operations/read-evidence-detail.ts
import { failure, ok } from "@asdl/clinkr";
import { z } from "zod";

import type { AretroCliContext } from "../context.ts";
import { PayloadError } from "../payloads/errors.ts";
import { readJsonPayloadArtifact, resolveJsonPointer, type JsonValue } from "../payloads/lookup.ts";

export const readEvidenceDetailRequestSchema = z.object({
	payloadPath: z.string(),
	jsonPointer: z.string(),
});

export type ReadEvidenceDetailRequest = z.infer<typeof readEvidenceDetailRequestSchema>;

export const readEvidenceDetailResultSchema = z.object({
	payload_path: z.string(),
	json_pointer: z.string(),
	value: z.unknown(),
});

export type ReadEvidenceDetailResult = z.infer<typeof readEvidenceDetailResultSchema>;

export async function runReadEvidenceDetail(
	_context: AretroCliContext,
	request: ReadEvidenceDetailRequest,
) {
	const pointerError = dataPointerError(request.jsonPointer);
	if (pointerError !== null) {
		return failure("invalid_request", pointerError);
	}

	let envelope: JsonValue;
	try {
		envelope = readRawPayloadEnvelope(request.payloadPath);
		validateSuccessEnvelope(envelope, request.payloadPath);
		validatePayloadData(envelope, request.payloadPath);
	} catch (error) {
		return failureForPayloadError(error);
	}

	let value: JsonValue;
	try {
		value = resolveDetailValue(envelope, request.jsonPointer);
	} catch (error) {
		return failureForPayloadError(error);
	}

	return ok({
		payload_path: request.payloadPath,
		json_pointer: request.jsonPointer,
		value,
	});
}

export function renderReadEvidenceDetail(result: ReadEvidenceDetailResult): string {
	return JSON.stringify(result, null, 2);
}

function dataPointerError(pointer: string): string | null {
	if (pointer === "/data" || pointer.startsWith("/data/")) {
		return null;
	}
	return `JSON Pointer must target the payload data document under /data: ${JSON.stringify(pointer)}`;
}

function readRawPayloadEnvelope(payloadPath: string): JsonValue {
	const document = readJsonPayloadArtifact(payloadPath, { allowedRoles: new Set(["raw"]) });
	if (typeof document !== "object" || document === null || Array.isArray(document)) {
		throw new PayloadError(
			"payload_lookup_failed",
			`Raw payload artifact must contain a Clinkr envelope object: ${payloadPath}`,
		);
	}
	return document;
}

function validateSuccessEnvelope(envelope: JsonValue, payloadPath: string): void {
	if (typeof envelope !== "object" || envelope === null || Array.isArray(envelope)) {
		throw new PayloadError("payload_lookup_failed", `Invalid envelope: ${payloadPath}`);
	}
	const exitCode = (envelope as Record<string, unknown>).exit_code;
	if (typeof exitCode !== "number" || exitCode !== 0) {
		throw new PayloadError(
			"payload_lookup_failed",
			`Raw payload artifact must be a successful Clinkr envelope: ${payloadPath}`,
		);
	}
	if (!("data" in (envelope as Record<string, unknown>))) {
		throw new PayloadError(
			"payload_lookup_failed",
			`Raw payload artifact is missing Clinkr data: ${payloadPath}`,
		);
	}
}

function validatePayloadData(envelope: JsonValue, payloadPath: string): void {
	if (typeof envelope !== "object" || envelope === null || Array.isArray(envelope)) {
		throw new PayloadError("payload_lookup_failed", `Invalid envelope: ${payloadPath}`);
	}
	const data = (envelope as Record<string, unknown>).data;
	if (typeof data !== "object" || data === null || Array.isArray(data)) {
		throw new PayloadError(
			"payload_lookup_failed",
			`Raw payload artifact data must be an aretro detail object: ${payloadPath}`,
		);
	}
	const schemaVersion = (data as Record<string, unknown>).schema_version;
	if (typeof schemaVersion !== "number") {
		throw new PayloadError(
			"payload_lookup_failed",
			`Raw payload artifact data is missing a supported schema version: ${payloadPath}`,
		);
	}
	if (schemaVersion !== 1) {
		throw new PayloadError(
			"payload_lookup_failed",
			`Raw payload artifact schema version is unsupported: ${payloadPath}`,
		);
	}
}

function resolveDetailValue(envelope: JsonValue, pointer: string): JsonValue {
	return resolveJsonPointer(envelope, pointer);
}

function failureForPayloadError(error: unknown) {
	const payloadError = error instanceof PayloadError ? error : null;
	return failure(
		payloadError?.errorType ?? "payload_lookup_failed",
		payloadError?.message ?? String(error),
	);
}
